lib/cdo/aws/s3.rb

```diff
@@ -56,28 +56,63 @@ def self.public_url(bucket, filename)
       Aws::S3::Object.new(bucket, filename, region: CDO.aws_region).public_url
     end
 
-    class PublicVersionedLogUploader
-      def initialize(bucket, prefix)
+    class LogUploader
+      # A LogUploader is constructed with some preconfigured settings that will
+      # apply to all log uploads - presumably you may be uploading many similar
+      # logs.
+      # @param [String] bucket name on S3
+      # @param [String] prefix to prepend to all log keys/filenames
+      # @param [Bool] make_public will cause the uploaded files to be given
+      #        public_read permission, and public URLs to be returned.  Otherwise,
+      #        files will be private and short-term presigned URLs will be returned.
+      # @param [Hash] options applied to all uploads - can be overwritten by
+      #        options in individual upload calls.
+      def initialize(bucket, prefix, make_public=false)
         @bucket = bucket
         @prefix = prefix
+        @make_public = make_public
       end
 
-      #
-      # Uploads the file with the given filename to S3 in the preconfigured
-      # bucket and prefix, and returns a public URL to the uploaded log file.
-      # May raise an exception if the file cannot be opened or the S3 upload fails.
-      #
-      def upload_log(filename)
-        File.open(filename, 'rb') do |file|
-          result = AWS::S3.create_client.put_object(
+      # Uploads a log to S3 at the given key (appended to the configured prefix),
+      # returning a URL to the uploaded file.
+      # @param [String] name where the log will be placed under the configured prefix
+      # @param [String] body of the log to be uploaded
+      # @param [Hash] options
+      # @return [String] public URL of uploaded log
+      # @raise [Exception] if the S3 upload fails
+      # @see http://docs.aws.amazon.com/sdkforruby/api/Aws/S3/Client.html#put_object-instance_method for supported options
+      def upload_log(name, body, options={})
+        key = "#{@prefix}/#{name}"
+        puts "The key is #{key}"
+
+        options[:acl] = 'public-read' if @make_public
+        result = AWS::S3.create_client.put_object(
+          options.merge(
             bucket: @bucket,
-            key: "#{@prefix}/#{filename}",
-            body: file,
-            acl: 'public-read'
+            key: key,
+            body: body
           )
-          log_url = "https://s3.amazonaws.com/#{S3_LOGS_BUCKET}/#{S3_LOGS_PREFIX}/#{filename}"
+        )
+        if @make_public
+          log_url = AWS::S3.public_url(@bucket, key)
           log_url += "?versionId=#{result[:version_id]}" unless result[:version_id].nil?
-          return log_url
+          log_url
+        else
+          options = {bucket: @bucket, key: key, expires_in: 60.minutes}
+          options[:version_id] = result[:version_id] unless result[:version_id].nil?
```

Comment:
Note: I'm not 100% sure that this will work, and have [opened a StackOverflow question](http://stackoverflow.com/questions/37714197/how-do-i-create-a-presigned-link-to-a-particular-version-of-an-object-using-the) looking for more information.  This code path isn't exercised in our current configuration anyway - our versioned logs (UI test logs) are public, and our private logs are in an unversioned bucket - but I wanted to give it a shot in case we change our minds.